### PR TITLE
GraphError Improvements

### LINKED ISSUE #312: GraphError improvements - Move to Stable
- Remove NoneError
- Possibly refactor GraphError to use a error crate like thiserror

### PR DESCRIPTION
**Pull request by ecclarke42**
Closes #312 

Simplify GraphError by using `thiserror` and remove `NoneErrror` handling so errors can be `stable` compatible.

### CODE REVIEW

**Comment by ecclarke42** [comment]
@sreeise, while wrapping up some of these I saw that in `graph-http::iotools` the `copy` fn uses a few `expect`/`unwrap`s that need some added `GraphFailure` variants. That led me to scan the repo for more, of which there are a few.

1. How do you feel about adding some more specialized error types? E.g. an error for joining the file copying thread should only ever get returned from `IoTools::copy` (which could probably just be a function instead of a method), which is only ever called by `BlockingDownload`. So a BlockingDownloadError could capture that error in one variant and GraphFailure in another, so that consumers of other functions wouldn't need to handle the irrelevant error case. There is more of the same in some other client methods.
2. I also noticed during this scan that `GraphError` and `GraphResponse` use a `u16` instead of `http::StatusCode` (and `GraphError` re-implements some of the error messages already there. Just passing the `StatusCode` directly could simplify some things, if you're ok with that.

**Comment by sreeise** [comment]
Looks good. Currently working on another branch that will fix the build errors in the test pipeline. When that branch is merged then the tests can be rerun on your PR and as the tests are good, I will merge it into master.

**Comment by sreeise** [comment]
Specialized error types would be great, but not necessary in this PR unless you really want to do it.

Using StatusCode makes sense to me. So, as long as your up for doing it, then yeah im on board.

**Comment by ecclarke42** [comment]
Yeah, I'll take a stab at some of those changes and let you know what I come up with!

**Comment by sreeise** [comment]
> Yeah, I'll take a stab at some of those changes and let you know what I come up with!

Also, I typically use rustfmt for formatting and clippy for static analysis before merging to master. If its possible, could you run `cargo fmt` to format and fix clippy lints that may have arrived from your changes by running `cargo clippy`. If there are lints that need to be fixed that are not part of your changes then that is my fault, you can leave those and I will go back and fix them.

**Comment by sreeise** [comment]
The tests are only failing because they can't access the Action secrets needed for calling the graph API. But they do pass when I run them.

File: graph-http/src/download/client.rs
Comment (sreeise) [inline graph-http/src/download/client.rs:196]: Can we move this and the async error struct to a new file: graph-http/src/download/error.rs

File: graph-http/src/download/client.rs
Comment (sreeise) [inline graph-http/src/download/client.rs:197]:
```suggestion
#[allow(clippy::large_enum_variant)]
pub enum BlockingDownloadError {
```
This is needed on both the blocking download and async download or clippy will throw a fit for a large enum variant. I'm not worried about this lint, at least for now, because it doesn't have a performance impact.

File: graph-http/src/download/client.rs
Comment (sreeise) [inline graph-http/src/download/client.rs:366]:
```suggestion
#[allow(clippy::large_enum_variant)]
pub enum AsyncDownloadError {
```
This is needed on both the blocking download and async download or clippy will throw a fit for a large enum variant. I'm not worried about this lint, at least for now, because it doesn't have a performance impact.

File: graph-http/src/iotools.rs
Comment (sreeise) [inline graph-http/src/iotools.rs:61]: I have some concerns about these errors that you have added. My main concern is that I am unsure what adding these will benefit the project or those that use it. Do they provide more info on the error? I can possibly see that having these may indicate what type of error it is but is that really any different than what was already in place? Just looking for some context here.

**Review by sreeise — CHANGES_REQUESTED** [review]

**Comment by ecclarke42** [comment]
I usually write in vscode with rustfmt and clippy on-save, so they should be pretty close. Running a `cargo +nightly fmt` creates a ton of changes due to import rearranging, so I thought I'd ask if you want those before committing? I'd assume so, since it should conform to the `rustfmt.toml` that you have in the repo (and `+nightly` since `imports_granularity` requires it).

File: graph-http/src/iotools.rs
Comment (ecclarke42) [reply graph-http/src/iotools.rs:61]: I think that's a fair criticism. It may be something of a philosophical difference/preference, and since you're the owner/maintainer of this tool, I'll go with whatever you prefer. Sorry in advance for the wall of text, but I've been thinking a lot about this recently in other contexts and wanted to get some thoughts down.

---

## Typestate

One of Rust's great strengths, imo, is the "correctness" aspect of the type system. You can construct your code such that invalid states are actually impossible. For example, `iotools::copy()` before returned a `GraphFailure` error, but it is impossible for it to generate a `serde_yaml::Error` or `handlebars::RenderError`. While it's cleaner in the library code, a consumer of the library would still need to match on all of those cases to either recover or panic.

Granted, in the case of `iotools`, an api consumer isn't going to use it directly, but more likely though one of the download clients. However, the story there is the same and we need to wrap the output of `iotools::copy` anyway to return a single error type.

Additionally, handling all the possible errors in `iotools::copy` without `panic`ing on threading errors with `GraphFailure` meant adding two new variants to `GraphFailure` which would never be relevant to any other error producer. Especially in a library, we probably want to let the api consumer decide when to `panic`/`expect`/`unwrap`. Searching through the library, there are a fair number of other panics that we could probably get rid of with a similar approach.

## Library Errors

A good example of the different approaches is in the difference between the target audiences of error crates, like [`anyhow`](https://github.com/dtolnay/anyhow) and [`thiserror`](https://github.com/dtolnay/thiserror) (both written by the same author, giving [this](https://github.com/dtolnay/thiserror#comparison-to-anyhow) comparison). `anyhow` is targeted at application errors, receiving just about any error type and presenting it to the user (and probably exiting without recovery). `thiserror` is more focused at nesting/combining error types for libraries, so that the consumer can decide what to do with them.

It feels like `GraphFailure` is acting a little more like an application error than a library error, and this is probably a first step at taking it the other direction. I don't think all of those changes will be part of this PR, but probably something to open for a future issue.

**Comment by sreeise** [comment]
> I usually write in vscode with rustfmt and clippy on-save, so they should be pretty close. Running a `cargo +nightly fmt` creates a ton of changes due to import rearranging, so I thought I'd ask if you want those before committing? I'd assume so, since it should conform to the `rustfmt.toml` that you have in the repo (and `+nightly` since `imports_granularity` requires it).

Is i`mports_granularity` not on stable? Maybe I misread, if its not on stable then it can just be removed.

File: graph-http/src/iotools.rs
Comment (sreeise) [reply graph-http/src/iotools.rs:61]: Im open to trying new things, and I like the way you presented it so lets keep it.

**Review by sreeise — APPROVED** [review]

**Comment by sreeise** [comment]
@ecclarke42 Removed the unstable parts in rustfmt as well as the feature attributes. Master is now using stable rust. Thanks for all the help. I will put an updated version on crates.io soon.

**Pull request merged**
